File: test_contest/solutions/knights.cpp
/**
 * Knights, Pirates, Ninjas
 * Author: Tyler Woodhull
 */

// Include the required headers
#include <stdio.h>
#include <string.h>

int main() {
   // Create a character array to store our number
   // We will need 10 characters for the number, plus an extra for \0
   char number[11];
   // Create variables to be used in the solution
   int n, i, j, numberLen;
   // Create flags that will store the answer of the case
   bool knights, pirates, ninjas;

   // Scan the number of test cases
   scanf("%d", &n);

   // Iterate through all of the test cases
   for (i = 0; i < n; i++) {
      // Initialize all answers
      knights = false;
      pirates = false;
      ninjas = false;

      // Scan in the number that will be evaluated
      scanf("%s", number);

      // Get the number of digits in the number
      numberLen = strlen(number);

      // Iterate through each digit
      for (j = 0; j < numberLen; j++) {
         // If the current digit is 0, then Ankit will think of ninjas
         if (number[j] == '0') {
            ninjas = true;
         }
         // If the current digit is 1, then Ankit will think of knights
         if (number[j] == '1') {
            knights = true;
         }
         // If the current digit is going to carry over when multiplied by 2
         // then Ankit will think of pirates
         if (number[j] >= '5') {
            pirates = true;
         }
      }

      // Print out the answers
      printf("%s", number);
      if (knights) printf(" knights");
      if (pirates) printf(" pirates");
      if (ninjas) printf(" ninjas");
      printf("\n");
   }

   return 0;
}

File: test_contest/solutions/knights.py
# Solution written by Will Cromar
# Python 3.6

import re

# The number of test cases
t = int(input())

# Run each test case
for _ in range(t):
    # We won't bother parsing an int from the input string
    num = input().strip()

    # List of strings to be printed at the end
    output = [num]

    # num reminds Ankit of knights if it contains a 1. Charge on!
    if re.search("1", num):
        output.append("knights")

    # num reminds Ankit of pirates if there is a digit that carrys
    # when doubled. (ie 5, 6, 7, 8, and 9)
    if re.search("[5-9]", num):
        output.append("pirates")

    # num reminds Ankit of ninjas if it contains a 0
    if re.search("0", num):
        output.append("ninjas")

    # Print all output tokens, separated by spaces
    print(" ".join(output))
